test/Validate.CFG.cpp

```diff
@@ -0,0 +1,601 @@
+}
+
+TEST_F(ValidateCFG, SwitchTargetFirstBlockBad) {
+  Block entry("entry");
+  Block bad("bad", SpvOpSwitch);
+  Block block1("block1");
+  Block block2("block2");
+  Block block3("block3");
+  Block def("def");  // default block
+  Block merge("merge");
+  Block end("end", SpvOpReturn);
+
+  bad.setBody(
+      "%cond    = OpSLessThan %intt %one %two\n"
+      "OpSelectionMerge %merge None\n");
+
+  string str = header + nameOps("entry", "bad", make_pair("func", "Main")) +
+               types_consts + "%func    = OpFunction %voidt None %funct\n";
+
+  str += entry >> bad;
+  str += bad >> vector<Block>({def, block1, block2, block3, entry});
+  str += def >> merge;
+  str += block1 >> merge;
+  str += block2 >> merge;
+  str += block3 >> merge;
+  str += merge >> end;
+  str += end;
+  str += "OpFunctionEnd\n";
+
+  CompileSuccessfully(str);
+  ASSERT_EQ(SPV_ERROR_INVALID_CFG, ValidateInstructions());
+  EXPECT_THAT(getDiagnosticString(), HasSubstr("First block"));
+  EXPECT_THAT(getDiagnosticString(), HasSubstr("entry"));
+  EXPECT_THAT(getDiagnosticString(), HasSubstr("Main"));
+  EXPECT_THAT(getDiagnosticString(), HasSubstr("bad"));
+}
+
+TEST_F(ValidateCFG, BranchToBlockInOtherFunctionBad) {
+  Block entry("entry");
+  Block middle("middle", SpvOpBranchConditional);
+  Block end("end", SpvOpReturn);
+
+  middle.setBody(
+      "%cond    = OpSLessThan %intt %one %two\n"
+      "OpSelectionMerge %end None\n");
+
+  Block entry2("entry2");
+  Block middle2("middle2");
+  Block end2("end2", SpvOpReturn);
+
+  string str = header + nameOps("middle2", make_pair("func", "Main")) + types_consts +
+               "%func    = OpFunction %voidt None %funct\n";
+
+  str += entry >> middle;
+  str += middle >> vector<Block>({end, middle2});
+  str += end;
+  str += "OpFunctionEnd\n";
+
+  str += "%func2    = OpFunction %voidt None %funct\n";
+  str += entry2 >> middle2;
+  str += middle2 >> end2;
+  str += end2;
+  str += "OpFunctionEnd\n";
+
+  CompileSuccessfully(str);
+  ASSERT_EQ(SPV_ERROR_INVALID_CFG, ValidateInstructions());
+  EXPECT_THAT(getDiagnosticString(), HasSubstr("Main"));
+  EXPECT_THAT(getDiagnosticString(), HasSubstr("middle2"));
+  EXPECT_THAT(getDiagnosticString(),
+              HasSubstr("are referenced but not defined"));
+}
+
+TEST_F(ValidateCFG, HeaderDoesntDominatesMergeBad) {
+  Block entry("entry");
+  Block bad("bad", SpvOpBranchConditional);
+  Block f("f");
+  Block merge("merge", SpvOpBranchConditional);
+  Block cont("cont");
+  Block end("end", SpvOpReturn);
+
+  bad.setBody("OpLoopMerge %merge %cont None\n");
+
+  merge.setBody(
+      " %cond    = OpSLessThan %intt %one %two\n"
+      "OpSelectionMerge %end None\n");
+
+  string str = header + nameOps("bad", "merge", make_pair("func", "Main")) +
+               types_consts + "%func    = OpFunction %voidt None %funct\n";
+
+  str += entry >> merge;
+  str += merge >> vector<Block>({bad, end});
+  str += bad >> vector<Block>({cont, f});
+  str += cont >> merge;
```

Comment:
This is a back edge not going into a loop header.  In the future, this (instead of merge domination) may trigger the validation failure and make this test fail despite the code working correctly.